### ISSUE TITLE
Production Hardening Sprint Lazy Datasets HF API Compatibility Trainer UX and Device Safety

### ISSUE THREAD

**Issue by david-thrower** [david-thrower/HelixLM#12]
### 🎯 Scope

This is a **single coordination issue** to track the systemic changes needed to move the HelixLM training stack from "research/demo" to production-ready. It subsumes eight internal bug reports and one documentation RFC. All changes must remain compatible with **Python 3.12** and the latest versions of `torch`/`transformers`/`datasets` specified in `requirements.txt`.

---

### 1. Dataset & DataLoader Layer (Issues #1, #2, #7)

**Problem:** 

- `DocumentAwareDataset` and `HelixDataset` eagerly tokenize all documents in `__init__`. For 715K documents this causes 5–20 minute silent hangs with zero GPU utilization.  
- `HelixDataset` inherits from `IterableDataset`, making `DataLoader(..., shuffle=True)` illegal in modern PyTorch (`ValueError`).  
- The pipeline materializes HF streams into Python strings instead of leveraging `datasets.IterableDataset` natively.

**Unified Fix:**  
- [ ] **Make lazy loading the default (or at least optional)** for all custom datasets. Store raw strings; call `tokenizer.encode()` inside `__iter__`/`__getitem__`. Chunking/rolling windows happen per-batch during iteration.  
- [ ] **Convert `HelixDataset` to a map-style `Dataset`** (index-based) so `DataLoader(shuffle=True)` works natively and integrates cleanly with HF `Trainer`.  
- [ ] **Implement `HelixHFDataset`** (or harden the existing one) to wrap `datasets.Dataset` / `IterableDataset` directly, supporting:
  - Streaming-first consumption (`load_dataset(..., streaming=True)`).
  - `max_samples` via `dataset.take(n)` or `dataset.select()` rather than Python list slicing.
  - On-the-fly tokenization and chunking.  
- [ ] Ensure all dataset constructors are **O(1)** and defer compute to the first batch iteration.

---

### 2. HuggingFace Hub Integration (Issues #4, #5)

**Problem:**  
- `DatasetDict.push_to_hub(..., commit_message="...")` raises `TypeError` in recent `datasets` versions.  
- `load_dataset(..., trust_remote_code=True)` is unconditionally passed to FineWeb-Edu, OpenWebMath, and OpenHermes, which are now standard Parquet datasets. On `datasets >= 3.0` this raises `ValueError`.

**Fix:**  
- [ ] Remove `commit_message` from all `push_to_hub` calls in `prepare_helixlm_dataset.py` (and audit `save_pretrained` calls for similar API drift).  
- [ ] Also `trust_remote_code=True` kwarg is no longer supported.

---

### 3. Trainer Hardening & UX (Issues #3, #6)

**Problem:**  
- Training scripts hardcode `use_titans_memory=True` in `HelixConfig`, and `Trainer` neither validates config/model agreement nor exposes clean dataset overrides.  
- `Trainer.train_epoch()` prints only every 10 batches, leaving users with 10–30 minutes of radio silence on slow CPU graph loops.

**Fix:**  
- [ ] Allow `Trainer.__init__` to accept optional `train_loader` / `val_loader` overrides so callers can inject lazy/streaming datasets.  
- [ ] Add `Trainer.validate_config()` that **warns** (does not error) if `use_titans_memory=True` is set on models with `< 50M` parameters and `seq_len < 512`, since Titans memory benefits are marginal at that scale.  
- [ ] **Propagate `use_titans_memory` as an overridable default** (default `True`, but allow scripts to set `False`).  
- [ ] Integrate `tqdm` into `Trainer` with throttled live updates (set postfix every batch, but log flush every 3–5 s to avoid spam). Add `verbose: bool = True` flag to `__init__` for headless runs.

---

### 4. Model / Device Safety (Issue #8)

**Problem:**  
- `HelixConfig.device="auto"` does not reliably move the model to CUDA during `from_pretrained()` or `__init__()`. Callers must manually call `.to(device)`, which is error-prone.

**Fix:**  
- [ ] In `HelixForCausalLM.__init__`, resolve `"auto"` explicitly: check `torch.cuda.is_available()` and emit a **warning** if it resolves to CPU despite CUDA being present.  
- [ ] Add a `.to_device()` convenience helper or respect `torch.set_default_device()` semantics.  
- [ ] Ensure `save_pretrained` / `from_pretrained` round-trips do not corrupt device placement metadata.

**Constraint:** Research and align with the **current supported device-initialization patterns** in the latest `transformers` and `torch` versions compatible with Python 3.12.

---

### 5. Developer Contract & Documentation (Meta)

**Problem:**  
There is no single source of truth for architectural standards.

**Fix:**  
- [ ] Create `ROADMAP.md` **and** `CONTRIBUTING.md` stating:
  - All dataset loading must support **streaming and lazy** execution.
  - All training scripts must show **progress** (`tqdm` or equivalent).
  - All HF API calls must be tested against the versions pinned in `requirements.txt`.
  - **No monkey-patching in user-facing examples**; if an internal API is broken, fix the library, not the demo script.
  - Target: production-ready AI training and fine-tuning code.

---

### 🏗 Proposed Unified Architecture (Reference)

```python
from datasets import load_dataset
from helix_lm import HelixTokenizer, HelixHFDataset, create_helix_dataloader, Trainer

# 1. Stream from Hub, never materialize full split
raw = load_dataset("HuggingFaceFW/fineweb-edu", "sample-10BT", 
                   split="train", streaming=True)

# 2. Lazy, map-style dataset
tokenizer = HelixTokenizer("gpt2")
train_ds = HelixHFDataset(raw, tokenizer, seq_len=256, max_samples=100_000)

# 3. Standard DataLoader with shuffle support
loader = create_helix_dataloader(train_ds, batch_size=4, shuffle=True)

# 4. Trainer with live progress and overridable config
trainer = Trainer(model=model, config=config, 
                  train_loader=loader, 
                  use_titans_memory=config.use_titans_memory,
                  verbose=True)
trainer.train()
```

---

### ✅ Acceptance Criteria (Checklist)

- [ ] `DocumentAwareDataset` initializes in < 1 s for 700 K documents (lazy mode).  
- [ ] `HelixDataset` is a map-style `Dataset` and works with `DataLoader(shuffle=True)`.  
- [ ] `prepare_helixlm_dataset.py` runs without `TypeError` on `push_to_hub` and `ValueError` on `load_dataset`.  
- [ ] `Trainer` displays a `tqdm` bar with live loss / tok/s metrics.  
- [ ] `Trainer` accepts injected dataloaders and warns on sub-optimal Titans memory configs.  
- [ ] `HelixForCausalLM("auto")` warns if CPU is selected while CUDA is available.  
- [ ] `ROADMAP.md` and `CONTRIBUTING.md` are merged to `main`.  
- [ ] All changes pass against `requirements.txt` pins on Python 3.12.

---

### 📎 Related Files to Touch
- `helix_lm/datasets.py` (`DocumentAwareDataset`, `HelixDataset`, `HelixHFDataset`)  
- `helix_lm/trainer.py` (`Trainer`, `create_helix_dataloader`)  
- `helix_lm/model.py` (`HelixForCausalLM`, `HelixConfig`, device logic)  
- `scripts/prepare_helixlm_dataset.py` (HF push/load calls)  
- `scripts/quick_demo_cpu_titans_mem.py` (and other training scripts, for config overrides)  
- `requirements.txt` (verify `datasets`, `transformers`, `torch` compatibility)

---

### Next Steps
1. **Assignee:** TBD — looking for a core contributor to own this sprint.  
2. **Branch naming:** `feat/production-hardening-lazy-datasets`  
3. **Suggested PR strategy:** One stacked PR per section (Dataset, HF Integration, Trainer, Model/Device, Docs) to keep review manageable, all tracked against this issue.

**Labels:** `enhancement`, `bug`, `performance`, `dataset`, `trainer`, `hf-integration`, `meta`, `good first issue` (for sub-tasks), `documentation`

**Comment by david-thrower** [comment]
Closed in #9
